### ISSUE TITLE
Xargo can not find component `rust-src`

### ISSUE THREAD

**Issue by Geemili** [japaric/xargo#85]
Following along with the [intermezzOs](https://intermezzos.github.io/book/creating-our-first-crate.html), I installed xargo. After installing Xargo, I ran the following command:

```bash
$ xargo build --release --target ./x86_64-unknown-metalomai-gnu.json
error: `rust-src` component not found. Run `rustup component add rust-src`.
$
```

Following the instructions in the error message, I installed the `rust-src` component.

```bash
$ rustup component add rust-src
info: downloading component 'rust-src'
 26.4 MiB /  26.4 MiB (100 %) 377.6 KiB/s ETA:   0 s
info: installing component 'rust-src'
warning: could not delete temp directory: /home/geemili/.multirust/tmp/zgpxf8_tm0kk8a72_dir
$ 
```

Then I tried running xargo again.

```bash
$ xargo build --release --target ./x86_64-unknown-metalomai-gnu.json
error: `rust-src` component not found. Run `rustup component add rust-src`.
$ # :(
```

***

Xargo finds the current sysroot using [this code][find-sysroot]. In bash, that looks like:

```bash
$ rustc --print sysroot
/home/geemili/.multirust/toolchains/nightly-x86_64-unknown-linux-gnu
$
```

Then it walks the directories to [find a file named "version", in a directory named rust.][find-version]. Translating to bash again:

```bash
$ find /home/geemili/.multirust/toolchains/nightly-x86_64-unknown-linux-gnu/ -name "version"
/home/geemili/.multirust/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/test/run-make/version
$ 
```

The "version" file is not in the "rust" folder, so the program fails.

***

I don't know why where version is was changed, but depending on this seems like a hack. Perhaps a better way could be found to find the location of `rust-src`? Maybe someone from who works on rustup could comment.

**Comment by japaric** [comment]
`version` is still there in this nightly version:

```
rustc 1.14.0-nightly (3f4408347 2016-10-27)
```

Will check with a more recent nightly.

cc @brson What's a reliable way to find the `rust-src` component within the sysroot. I was under the impression that the packaging code was explicitly adding that `version` file and that I could rely on it but perhaps I was wrong. Should I search for the `libstd`/`libcore` directory inside the sysroot instead? (There's more than one hit for that search query tough) I recall reading somewhere that I should *not* assume that `rust-src` is always going to be in `$sysroot/lib/rustlib/src/rust`.

**Comment by japaric** [comment]
> Will check with a more recent nightly.

FWIW,

```
rustc 1.14.0-nightly (5665bdf3e 2016-11-02)
```

also has the file.

**Comment by Geemili** [comment]
Hmm. I'll try updating my version of rustc.

**Comment by Geemili** [comment]
Updating my version of rustc fixed the issue. Still, I would like to know if there is a better way to get the path to rust-src.

**Comment by k0pernicus** [comment]
@Geemili I think that you can't get the path to `rust-src` easily using `rustup`...
You can check if `rust-src` is installed using `rustup component list | grep rust-src` - and your current Rust version/compiler using `rustc --print sysroot`.
By default, `rustup` install source files/directories in `$(rustc --print sysroot)/lib/rustlib/src/`

**Comment by k0pernicus** [comment]
@japaric: I think that you can close this thread.

**Comment by jdub** [comment]
Looks like this has re-emerged with rustup 1.0.0 (using xargo master and Rust nightly-x86_64-unknown-linux-gnu 8f02c429a), quite possibly for different reasons.

Note that rustup 1.0.0 migrates `~/.multirust` to `~/.rustup` and replaces it with a symlink.

Just about to look at a strace to see if something obvious sticks out...

**Comment by jdub** [comment]
Hmm:

```rust
/// The `rust-src` component within `rustc`'s sysroot
pub fn rust_src() -> Result<PathBuf> {
    for entry in WalkDir::new(try!(sysroot())) {
        let entry =
            try!(entry.chain_err(|| "error recursively walking the sysroot"));

        if entry.file_type().is_file() && entry.file_name() == "version" {
            let path = entry.path();
```

Looks like an assumption changed:

```
$ find $(rustc --print sysroot) -name version -type f | wc -l
0
```

Is the equivalent of `$(rustc --print sysroot)/lib/rustlib/src/rust` too naive for xargo?

**Comment by jdub** [comment]
This is silly and I've removed the nice error messages, but it works. In `src/rustc.rs`.

```rust
/// The `rust-src` component within `rustc`'s sysroot
pub fn rust_src() -> Result<PathBuf> {
    Ok(try!(sysroot()).join("lib/rustlib/src/rust"))
}
```

**Comment by japaric** [comment]
@brson recommends changing the logic to search for the 'Cargo.toml' file that belongs to the `std` crate. [See this convo](https://botbot.me/mozilla/rust-tools/2016-12-17/?msg=78115688&page=1). If anyone wants to update logic, let me know. Otherwise, I'll look at it sometime during the weekend.

**Comment by jdub** [comment]
Taking a run at this now.

**Comment by thejpster** [comment]
Is there a workaround? I was hoping to get some work done this weekend, but I get the 'rust-src component not found' with the latest nightly.

**Comment by jdub** [comment]
Just use the PR above. 😃

**Comment by thejpster** [comment]
Doesn't work for me.

manifest path `/home/jonathan/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/src/libstd/Cargo.toml` does not exist.

**Comment by jdub** [comment]
Hmm, double `src` is interesting, I'll have to fix that. But it shouldn't even get there if you're using rustup. Do you have the `rust-src` component installed? `rustup component add rust-src`.

**Comment by thejpster** [comment]
I do.

**Comment by jdub** [comment]
Oh, that's coming from `src/dag.rs`. It's expecting rust_src to just be `<...>/rust`. I wonder why this is even working on my system...

**Comment by jdub** [comment]
Guess who never cleared his xargo build cache? How embarrassment. One moment, I'll have this fixed in a jiffy. 😄

**Comment by jdub** [comment]
OK, sorry about that -- give it a whirl now. I suspect the fallback won't work yet, but your default rustup environment should be fine.

**Comment by thejpster** [comment]
I think I've fixed it to :)

**Comment by TheZoq2** [comment]
@thejpster Your fix seems to have fixed it for me aswell.

**Comment by thejpster** [comment]
@japaric - will there be a release soon? My travis build uses Xargo on crates.io, not my funky patched version.

**Comment by japaric** [comment]
@thejpster Just released 0.2.3.
